### ISSUE TITLE
Exception when writing partial response with expression yielding null

### ISSUE THREAD

**Issue by arjantijms** [eclipse-ee4j/mojarra#4925]
When writing a partial response, e.g.

```text
foo bar kar #{somebean.foo} foo bar
```

And `#{somebean.foo}` resolves to a null (that gets coerced to "") or directly returns a "" then the following exception is thrown:

```
java.lang.ArrayIndexOutOfBoundsException: Index -2 out of bounds for length 0
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeUnescapedCData(HtmlResponseWriter.java:1474)
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeText(HtmlResponseWriter.java:961)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:296)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:296)
	at org.richfaces.context.PartialResponseWriterWrapper.writeText(PartialResponseWriterWrapper.java:209)
```

The case for length == 0 seems to be missing here:

```java
private void writeUnescapedCData(char[] cbuf, int offset, int length) throws IOException {
        // single char case
        if (length == 1) {
            if (cbuf[offset] == ']') {
                appendBuffer(ESCAPEDSINGLEBRACKET);
            } else {
                appendBuffer(cbuf[offset]);
            }
            flushBuffer();
            return;
        }
        // two char case
        if (length == 2) {
            if (cbuf[offset] == ']' && cbuf[offset + 1] == ']') {
                appendBuffer(ESCAPEDSINGLEBRACKET);
                appendBuffer(ESCAPEDSINGLEBRACKET);
            } else {
                appendBuffer(cbuf[offset]);
                appendBuffer(cbuf[offset + 1]);
            }
            flushBuffer();
            return;
        }
        // > 2 char case
        boolean last = false;
        for (int i = offset; i < length - 2; i++) {
            if (cbuf[i] == ']' && cbuf[i + 1] == ']' && cbuf[i + 2] == '>') {
                appendBuffer(ESCAPEDEND);
                i += 2;
            } else {
                appendBuffer(cbuf[i]);
            }
            if (i == (offset + length - 1)) {
                last = true;
            }
        }
        // if we didn't look at the last characters, look at them now
        if (!last) {
            if (cbuf[offset + length - 2] == ']') {
                appendBuffer(ESCAPEDSINGLEBRACKET);
            } else {
                appendBuffer(cbuf[offset + length - 2]);
            }
            if (cbuf[offset + length - 1] == ']') {
                appendBuffer(ESCAPEDSINGLEBRACKET);
            } else {
                appendBuffer(cbuf[offset + length - 1]);
            }
        }
        flushBuffer();
}
```

The method was added here: https://github.com/eclipse-ee4j/mojarra/commit/57261962e6f9250b33bbe790c0f1a0b304ea5013 by @TomasHofman although it was a long time ago, and I haven't seen other rapports until now.

@TomasHofman what do you think? I intend to just handle the 0 length case with a return.

@BalusC any input, preferences here?

**Comment by TomasHofman** [comment]
@arjantijms I agree, thank you for fixing this!

**Comment by BalusC** [comment]
LGTM!

Only missing unit test.

**Comment by arjantijms** [comment]
> Only missing unit test.

Okay, I'll look for a test

**Comment by melloware** [comment]
@arjantijms I wonder if that is the same issue found here: https://github.com/primefaces/primefaces/issues/7343

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 6
    at javax.faces.component.UIComponentBase.processRestoreState (UIComponentBase.java:1399)
    at javax.faces.component.UIComponentBase.processRestoreState (UIComponentBase.java:1420)
    at javax.faces.component.UIViewRoot.processRestoreState (UIViewRoot.java:875)
    at com.sun.faces.application.view.JspStateManagementStrategy.restoreView (JspStateManagementStrategy.java:273)
    at com.sun.faces.application.StateManagerImpl.restoreView (StateManagerImpl.java:138)
    at com.sun.faces.application.view.ViewHandlingStrategy.restoreView (ViewHandlingStrategy.java:123)
    at com.sun.faces.application.view.FaceletViewHandlingStrategy.restoreView (FaceletViewHandlingStrategy.java:589)
    at com.sun.faces.application.view.MultiViewHandler.restoreView (MultiViewHandler.java:151)
    at javax.faces.application.ViewHandlerWrapper.restoreView (ViewHandlerWrapper.java:353)
    at javax.faces.application.ViewHandlerWrapper.restoreView (ViewHandlerWrapper.java:353)
    at com.sun.faces.lifecycle.RestoreViewPhase.execute (RestoreViewPhase.java:199)
    at com.sun.faces.lifecycle.Phase.doPhase (Phase.java:101)
    at com.sun.faces.lifecycle.RestoreViewPhase.doPhase (RestoreViewPhase.java:123)
    at com.sun.faces.lifecycle.LifecycleImpl.execute (LifecycleImpl.java:198)
```

Reproducer:
[pf-7343.zip](https://github.com/primefaces/primefaces/files/6528034/pf-7343.zip)

It looks like a Mojarra bug as it works fine in Myfaces.

`mvn clean jetty:run -Pmyfaces22` - WORKS

`mvn clean jetty:run -Pmyfaces23` - WORKS

`mvn clean jetty:run -Pmojarra22` - **ERROR**

`mvn clean jetty:run -Pmojarra23` - **ERROR**

**Comment by melloware** [comment]
Actually this doesn't look related I will open a new ticket

**Comment by BalusC** [comment]
Merged and upmerged in 3.0 and 4.0 as well.

https://github.com/eclipse-ee4j/mojarra/commit/77e0e04a3027c73bc62383a2ac3cde3b86fe3f32
https://github.com/eclipse-ee4j/mojarra/commit/7ebc063432f58f8a5fb0d9fea565f46ed44a29b5
